cell_classification/predict.py

```diff
@@ -0,0 +1,236 @@
+
+
+def load_model_and_val_data(params):
+    """Load model and validation data from params dict
+    Args:
+        params (dict):
+            dictionary containing model and validation data paths
+    Returns:
+        model (ModelBuilder):
+            trained model
+        val_data (tf.data.Dataset):
+            validation dataset
+    """
+    params["eval"] = True
+    model = ModelBuilder(params)
+    model.prep_data()
+    model.load_model(params["model_path"])
+    val_dset = iter(model.validation_dataset)
+    return model, val_dset
+
+
+def predict(model, val_dset, params):
+    """Predict labels for validation data
+    Args:
+        model (ModelBuilder):
+            trained model
+        val_dset (tf.data.Dataset):
+            validation data
+        params (dict):
+            dictionary containing model and validation data paths
+    Returns:
+        predictions (list):
+            list of dictionaries containing the full example + prediction
+        params (dict):
+            dictionary containing model hyperparams and validation data paths
+    """
+    pred_list = []
+    for sample in tqdm(val_dset):
+        sample["prediction"] = model.predict(model.prep_batches(sample)[0])
+        pred_list.append(sample)
+
+    # split batches to single samples
+    single_example_list = []
+    for sample in pred_list:
+        for key in sample.keys():
+            sample[key] = np.split(sample[key], sample[key].shape[0])
+        for i in range(len(sample["prediction"])):
+            single_example = {}
+            for key in sample.keys():
+                single_example[key] = np.squeeze(sample[key][i], axis=0)
+                if single_example[key].dtype == object:
+                    single_example[key] = sample[key][i].item().decode("utf-8")
+            single_example_list.append(single_example)
+    # save pred_list to pickle file
+    params["eval_dir"] = os.path.join(model.params["model_dir"], "eval")
+    os.makedirs(params["eval_dir"], exist_ok=True)
+    with open(os.path.join(params["eval_dir"], "pred_list.pkl"), "wb") as f:
+        pickle.dump(single_example_list, f)
+
+    # save params to toml file
+    with open(os.path.join(params["model_dir"], "params.toml"), "w") as f:
+        toml.dump(params, f)
+    return single_example_list, params
+
+
+def calc_roc(pred_list, gt_key="marker_activity_mask", pred_key="prediction"):
+    """Calculate ROC curve
+    Args:
+        pred_list (list):
+            list of samples with predictions
+        gt_key (str):
+            key for ground truth labels
+        pred_key (str):
+            key for predictions
+    Returns:
+        roc (dict):
+            dictionary containing ROC curve data
+    """
+    roc = {"fpr": [], "tpr": [], "thresholds": [], "auc": []}
+    for sample in pred_list:
+        if sample[gt_key].max() == 0:
+            continue
+        fpr, tpr, thresholds = roc_curve(sample[gt_key].flatten(), sample[pred_key].flatten())
```

Comment:
This will need to be tweaked for cell-level metrics. Do you think the best thing is just get the pixel-level version working first, and then worry about cell-level in a different PR? Or try and do both here?

Reply:
I think small stuff like the above should be changed in this PR, but I'd like to do cell level stuff in a another PR since this one contains already a lot of changes.

Reply:
Sounds good!

Reply:
I'll leave the `.flatten()` here as it is, since it does nothing if the array is already flat which will be the case for cell-level data.